### PR TITLE
feat(M03): notebook train_numpy_nn — MLP NumPy/PyArrow e split T021

### LINKED ISSUE #14: T033 — Modelo de rede neural
```yaml
id: T033
story: S03
milestone: M03
status: Todo
depends_on: [T030]
artifacts:
  - notebook ou script da rede (Keras/PyTorch/sklearn MLP)
```

## Objetivo

Treinar uma **rede neural** para o mesmo alvo, com arquitetura e treino **documentados** (camadas, ativações, épocas, batch, early stopping se houver).

## Checklist

- [ ] Diagrama ou lista de camadas e unidades.
- [ ] Função de perda alinhada à tarefa.
- [ ] Registrar tempo aproximado de treino e hardware.
- [ ] Métricas no val/teste no protocolo de **T030**.

## Evidence

- Curva de loss opcional; métricas finais obrigatórias.

## Links

- Story: [../storys/S03-parte3-modelos-avaliacao.md](../storys/S03-parte3-modelos-avaliacao.md)
- Milestone: [../milestones/M03-modelagem-avaliacao.md](../milestones/M03-modelagem-avaliacao.md)

### PR DESCRIPTION
**Pull request by TalesAugusto1**
## Resumo

Introduz / mantém o notebook **[`notebooks/neural_network_numpy_training.ipynb`](notebooks/neural_network_numpy_training.ipynb)** (antes `train_numpy_nn.ipynb`) com **MLP em NumPy** + leitura em batches **PyArrow**, alinhado ao **split temporal T021** (70/15/15), resolução robusta de **`data/Indian_Weather_Dataset.parquet`**, normalização só no treino e correções que evitam **`IndexError`** nos índices one-hot, **`RuntimeWarning`** (média em slice vazio) e **loss não finita**. O modelo está alinhado ao protocolo **T030**: alvo **`temperature_C`**, perda **MSE**, saída **linear**, métricas **MAE**, **RMSE** e **R²** em val/teste (sem **BCE**/classificação). Inclui **atualização de biases** no `backward` do MLP, **`RNG` fixo**, erros explícitos se não houver batches, **`.gitignore`** para **`.ipynb_checkpoints/`**, documentação em **changelog** + evidência em **T033**.

## Motivação

- Entregável para **M03 / T033** (rede neural) reprodutível no host, sem Keras/PyTorch obrigatórios.
- Mesmo critério temporal que **`scripts/split_temporal.py`**, evitando treino vazio ou inconsistente com datas fixas.
- Notebook utilizável com **cwd** na raiz do repo ou em **`notebooks/`**.
- **Alvo e métricas** alinhados a **T030** (`temperature_C`, MAE/RMSE/R²), comparáveis ao baseline e ao `decision_tree.ipynb`.

## O que mudou

| Ficheiro | Alteração |
|----------|-----------|
| `notebooks/neural_network_numpy_training.ipynb` | `resolve_data_parquet()`, `temporal_split_filters()`, `get_stats()`, `to_xy` com índices one-hot estáveis e `nan_to_num`; **regressão**: `TARGET = temperature_C`, **`temperature_C` removido das features**, **`mse()`**, saída linear, **`evaluate()` → MAE, RMSE, R²**; MLP com biases no `backward`; renomeação do ficheiro para nome explícito de rede neural. |
| `docs/changelog/2026-04-28-m03-notebook-train-numpy-nn.md` | Changelog da entrega inicial (BCE / robustez); follow-up T030 marcado como coberto pela entrada seguinte. |
| `docs/changelog/2026-05-02-m03-t033-neural-network-temperature-regression.md` | Changelog da migração para **regressão em `temperature_C`** (T030). |
| `docs/changelog/README.md` | Índice (entradas **2026-05-02** + **2026-04-28**). |
| `storyline/tasks/T033-rede-neural.md` | **Evidence** com ligações ao notebook e aos dois changelogs (histórico + regressão). |
| `.gitignore` | Ignorar **`.ipynb_checkpoints/`**. |

*(Remove a linha sobre `notebooks/eda.ipynb` do PR se não estiver mesmo no diff deste branch.)*

## Como verificar

1. `Indian_Weather_Dataset.parquet` em **`data/`** na raiz do repo.  
2. Abrir **`notebooks/neural_network_numpy_training.ipynb`**, kernel com `pyarrow`, `numpy`, `matplotlib`.  
3. Correr células em sequência: **`Split T021: ... | train rows: ...`** com contagem **> 0**; por época **`mse=...`** finito; no fim **`VAL` / `TEST`** com **(MAE, RMSE, R²)**.  
4. (Opcional) `python scripts\split_temporal.py -i data\Indian_Weather_Dataset.parquet` para confrontar fronteiras temporais.

## Changelogs

- [docs/changelog/2026-05-02-m03-t033-neural-network-temperature-regression.md](https://github.com/TalesAugusto1/Big_Data-Indian_Weather/blob/trainnumpy/docs/changelog/2026-05-02-m03-t033-neural-network-temperature-regression.md) — regressão **`temperature_C`** / T030  
- [docs/changelog/2026-04-28-m03-notebook-train-numpy-nn.md](https://github.com/TalesAugusto1/Big_Data-Indian_Weather/blob/trainnumpy/docs/changelog/2026-04-28-m03-notebook-train-numpy-nn.md) — entrega inicial NumPy + PyArrow / T021  

*(Ajusta `trainnumpy` → branch real se for outro.)*

## Storyline / milestone

- **Milestone:** M03 — Modelagem e avaliação  
- **Tarefas:** **T033** (rede neural); **T021** (split temporal); **T030** (métricas de regressão no notebook).

{FDA1DA29-04AE-4100-99DB-EDEEC5A96828}
{5DF13FC7-2F8D-4250-B3A6-A6CAF07EE3B7}

### CODE REVIEW

File: notebooks/train_numpy_nn.ipynb [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+        "\n",
+        "# Mesmas frações que scripts/split_temporal.py (T021)\n",
+        "TRAIN_FRAC = 0.70\n",
+        "VAL_FRAC = 0.15\n",
+        "\n",
+        "BATCH_SIZE = 8192\n",
+        "EPOCHS = 3\n",
+        "LR = 0.001\n",
+        "\n",
+        "TARGET = \"rain_label\"\n",
```
Comment (folena) [inline notebooks/train_numpy_nn.ipynb]: Aqui o target tem que ser a temperatura, Troca pra coluna de temperatura e avalie se o modelo precisa de alterações, por se tratar de um modelo que foi treinado pra classificação em vez de regressão.

**Review by folena — CHANGES_REQUESTED** [review]

**Review by folena — APPROVED** [review]
